cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,282 @@
+  virtual Status Compute(FunctionContext* ctx, const ArrayData& left,
+                         const ArrayData& right, std::shared_ptr<ArrayData>& out) = 0;
+
+ public:
+  Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+              Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, right.kind());
+    DCHECK_EQ(Datum::ARRAY, left.kind());
+
+    const ArrayData& left_data = *left.array();
+    const ArrayData& right_data = *right.array();
+
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(Compute(ctx, left_data, right_data, result));
+
+    out->value = std::move(result);
+    return Status::OK();
+  }
+  std::shared_ptr<DataType> out_type() const override { return boolean(); }
+};
+
+// ----------------------------------------------------------------------
+// Using a visitor create a memo_table_ for the right array
+// TODO: Implement for small lists
+
+template <typename Type, typename Scalar>
+struct MemoTableRight {
+  Status VisitNull() { return Status::OK(); }
+
+  Status VisitValue(const Scalar& value) {
+    memo_table_->GetOrInsert(value);
+    return Status::OK();
+  }
+
+  Status Append(const ArrayData& right) {
+    memo_table_.reset(new MemoTable(0));
+    return ArrayDataVisitor<Type>::Visit(right, this);
+  }
+
+  using MemoTable = typename HashTraits<Type>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+};
+
+// ----------------------------------------------------------------------
+// Check if value in both arrays or not and returns boolean values
+
+// \brief Iterate over the left array using another visitor.
+// In VisitValue, use the memo_table_ (for right array) and check if value
+// in left array is in the memo_table_. Append the true to the
+// BooleanBuilder if condition satisfied, else false.
+
+// Additional member "right_null_count" is used to check if
+// null count in right is not 0, then append true to the BooleanBuilder
+// when left array has a null, else append false.
+
+template <typename T, typename Scalar>
+class IsInKernel : public HashBinaryKernelImpl {
+ public:
+  IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool) {}
+
+  Status VisitNull() {
+    if (right_null_count != 0) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppend(false);
+    }
+    return Status::OK();
+  }
+
+  Status VisitValue(const Scalar& value) {
+    if (memo_table_->Get(value) != -1) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppend(false);
+    }
+    return Status::OK();
+  }
+
+  Status Compute(FunctionContext* ctx, const ArrayData& left, const ArrayData& right,
+                 std::shared_ptr<ArrayData>& out) override {
+    bool_builder_.Reset();
+    bool_builder_.Reserve(left.length);
+    memo_table_.reset(new MemoTable(0));
+
+    MemoTableRight<T, Scalar> func;
+    func.Append(right);
+    memo_table_ = std::move(func.memo_table_);
+
+    right_null_count = right.null_count;
```

Comment:
On ArrayData, you need to invoke `GetNullCount()` since it's memoized.

Reply:
Changed it from null_count to GetNullCount().